### DISCUSSION TITLE
Beta test: v2.0

### DISCUSSION THREAD

**Discussion by samizdatco** [samizdatco/skia-canvas#189] [Announcements]
Hi all, I've posted a pre-release version of Skia Canvas 2.0 and would love some help finding any remaining show-stoppers before publishing it to NPM. If you wouldn't mind giving it a quick test, you can install the new version by typing:
> ```bash
> npm install https://github.com/samizdatco/skia-canvas
> ```

Please post general 👍/👎 experiences here or feel free to [open an issue](https://github.com/samizdatco/skia-canvas/issues) if you want to report something specific.

Thanks!

---
### New Features

- **Docs Website**
    - Documentation is now hosted at [skia-canvas.org](https://skia-canvas.org).
- **Imagery**
    - SVGs & WEBP bitmaps can now be loaded as Images and drawn to the canvas
    - Canvas [saveAs][Canvas.saveAs] and [toBuffer][Canvas.toBuffer] added support for WEBPs and ‘raw’ bitmaps of [different types][colorType]
    - [drawImage][mdn_drawImage] and [createPattern][mdn_createPattern] now work with [ImageData][ImageData] objects as well as [Images][Image] and [Canvases][Canvas]
    - New [loadImageData][loadimagedata] helper function to asynchronously load files & urls for direct pixel access
    - Image objects now support standard [on/off/once][img_bind] EventEmitter methods
- **Typography**
    - Added support for WOFF & WOFF2 font formats
    - Added generic font mappings for `sans-serif`, `serif`, `monospace`, and `system-ui`
    - New Context type properties: [textDecoration][textDecoration], [fontStretch][fontStretch], [letterSpacing][letterSpacing], & [wordSpacing][wordSpacing]
- **GUI**
    - New Window property [`resizable`][resizable] allows resizing to be disabled
    - Added keyboard support for CJK IME via [`compose*`][compositionstart] and [`input`][input] events
- **Rendering**
    - New Canvas [`engine`][engine] property gives status of GPU or CPU renderer
    - The number of background threads used for asynchronous exports can now be set with the [`SKIA_CANVAS_THREADS`][multithreading] environment variable
    - The sample-count used by the GPU for multiscale antialiasing can now be configured through the optional [`msaa`][msaa] export argument. If omitted, defaults to 4x MSAA.

### Breaking Changes
- The minimum required Node version is now 12.22+, 14.17+, or 16+.
- Images now load asynchronously in cases where the `src` property has been set to a local path. As a result, it's now necessary to `await img.decode()` or set up an `.on("load", …)` handler before drawing it—even when the `src` is non-remote.
- The **KeyboardEvent** object returned by the `keyup`/`keydown` and `input` event listeners now has fields and values consistent with browser behavior. In particular, `code` is now a name (e.g., `ShiftLeft` or `KeyS`) rather than a numeric scancode, `key` is a straightforward label for the key (e.g., `Shift` or `s`) and the new [`location`][key_location] field provides a numeric description of which variant of a key was pressed.
- The deprecated `.async` property has been removed. See the [v0.9.28](#--v0928--jan-12-2022) release notes for details.
- The non-standard `.textTracking` property has been removed in favor of the new [`.letterSpacing`][letterSpacing] property

### Bugfixes
- Made Vulkan device selection & initialization more reliable
- The **Image**.onload callback now properly sets `this` to point to the new image (contributed by @mpaperno & @ForkKILLET).
- Creating a **Window** with `fullscreen` set to `true` now takes effect immediately
- Drawing paths after setting an invalid transform no longer crashes (contributed by @mpaperno)
- Windows with `.on("draw")` handlers no longer [become unresponsive](https://github.com/gfx-rs/gfx/issues/2460) on macOS 14+ after being fully occluded by other windows
- Ellipses with certain combinations of positive and negative start- and stop-angles now render correctly—previously they would not appear at all if the total sweep exceeded 360° (contributed by @mpaperno)
- The `drawCanvas()` method now clips to the specified crop size (contributed by @mpaperno)
- [`isPointInPath`][isPointInPath()] and [`isPointInStroke`][isPointInStroke()] now work with **Path2D** objects

### Misc. Improvements
- Upgraded Skia to [milestone 131](https://github.com/rust-skia/rust-skia/releases/tag/0.80.0)
- Added more thorough TypeScript definitions
- Significant performance improvements to FontLibrary and [getImageData()][mdn_getImageData], which now makes use of the GPU (if enabled) and caches data between calls
- New Vulkan renderer using [Vulkano][vulkano] for window-drawing on Windows and Linux.
- Wait to initialize GPU until it is used, not at startup. Set Canvas[`.gpu`][canvas_gpu] to `false` to prevent any GPU-related resource acquisition from occurring
- Now supports ES Modules and the `import` statement
- The [getImageData()][mdn_getImageData] method now makes use of the GPU (if enabled) and caches data between calls, greatly improving performance for sequential queries

[resizable]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/window.md#resizable
[key_location]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/location
[pragmatrix_emergent]: https://github.com/pragmatrix/emergent/blob/master/src/skia_renderer.rs
[woff_wiki]: https://en.wikipedia.org/wiki/Web_Open_Font_Format
[css_transform]: https://developer.mozilla.org/en-US/docs/Web/CSS/transform
[DOMMatrix]: https://developer.mozilla.org/en-US/docs/Web/API/DOMMatrix
[FontLibrary.use]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/font-library.md#use
[Canvas.saveAs]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md#saveas
[Canvas.toBuffer]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md#tobuffer
[letterSpacing]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/letterSpacing
[wordSpacing]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/wordSpacing
[fontStretch]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fontStretch
[isPointInPath()]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/isPointInPath
[isPointInStroke()]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/isPointInStroke
[node_napi]: https://nodejs.org/api/n-api.html#node-api-version-matrix
[neon_rs]: https://neon-rs.dev
[msaa]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md#msaa
[multithreading]: https://github.com/samizdatco/skia-canvas/tree/main/docs/getting-started.md#multithreading
[compositionstart]: https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event
[compositionupdate]: https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionupdate_event
[compositionend]: https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionend_event
[input]: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event
[win_bind]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/window.md#on--off--once
[img_bind]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/image.md#on--off--once
[Canvas]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md
[Image]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/image.md
[ImageData]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/imagedata.md
[colorType]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/imagedata.md#colortype
[mdn_createPattern]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createPattern
[mdn_getImageData]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData
[mdn_createImageData]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createImageData
[mdn_putImageData]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData
[mdn_drawImage]: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage
[outline_text]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/context.md#outlinetext
[canvas_gpu]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md#gpu
[window]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/window.md
[loadimagedata]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/imagedata.md#loadimagedata
[textDecoration]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/context.md#textdecoration
[engine]: https://github.com/samizdatco/skia-canvas/tree/main/docs/api/canvas.md#engine
[vulkano]: https://github.com/vulkano-rs/vulkano

**Comment by loveloki** [comment]
Use `npm install https://github.com/samizdatco/skia-canvas` insatll,

And get this wrong:

![image](https://github.com/user-attachments/assets/c9672898-88bc-49d4-b0f9-3d72e80ba4de)

**Reply by loveloki** [reply]
fix pr see #191

**Reply by samizdatco** [reply]
I think this is now fixed as of a48726a060aad24ef47bf13d8f0343c5579e61a4. Can anyone confirm that both named and default `import`s are now working properly in the beta?

**Comment by MarosPistej** [comment]
I noticed two changes after upgrade from 1.0.2 to 2.0.0 

1. the font weight on vertical text render differently
2. some of texts (chart Y Axis labels) shifted, here i suspect that some change is in baseLine: middle, other texts which didn't moved have different baseLine

https://github.com/user-attachments/assets/56f26cc4-c641-42c9-a562-f252125728e5

Were there any changes in this area?
